system/include/emscripten/heap.h

```diff
@@ -13,10 +13,27 @@
 extern "C" {
 #endif
 
+// Returns a pointer to a memory location that contains the heap DYNAMICTOP
+// variable (the end of the dynamic memory region)
 intptr_t *emscripten_get_sbrk_ptr(void);
-int emscripten_resize_heap(size_t requested_size);
+
+// Attempts to geometrically or linearly increase the heap so that it
+// grows by at least requested_growth_bytes new bytes. The heap size may
+// be overallocated, see src/settings.js variables MEMORY_GROWTH_GEOMETRIC_STEP,
+// MEMORY_GROWTH_GEOMETRIC_CAP and MEMORY_GROWTH_LINEAR_STEP. This function
+// cannot be used to shrink the size of the heap.
+int emscripten_resize_heap(size_t requested_growth_bytes);
+
+// Returns the current size of the WebAssembly heap.
 size_t emscripten_get_heap_size(void);
 
+// Resizes the heap to the given size. This function cannot shrink the size,
+// so the size that is passed must be >= current emscripten_get_heap_size().
+// Returns 1 on success, 0 otherwise. In order to call this function,
+// the page must have been compiled with -s ALLOW_MEMORY_GROWTH=1 linker
+// flag.
+int emscripten_realloc_buffer(size_t requested_size);
```

Comment:
I'm unclear about heap growth vs growing the underlying wasm memory.  Are they always the same thing?

Does `emscripten_resize_heap` above also require `ALLOW_MEMORY_GROWTH`?  I'm a little confused a bout the use of the term "realloc" which normally implies the buffer could be moved.  Maybe there could be a better name for this function?  Is it basically like `emscripten_resize_heap` bug taking an absolute size?

Reply:
Since this function doesn't seem very well named perhaps we can avoid exposing it and keep it as an implementation detail.  I see its only used in test code anyway.  Could the test be modified to not depend on the this currently internal function?

Reply:
The intent is to be able to use `emscripten_realloc_buffer` in public code to resize the heap. In JS code we can use that to actually shrink the heap and release memory back to the browser.

`emscripten_resize_heap` only grows the heap with geometric/linear overgrowth, suitable for sbrk()-like allocators. `emscripten_realloc_buffer` is suitable for someone who is implementing their own `malloc()` that does not use `sbrk()`.

Reply:
The current  emscripten_resize_heap seems to take an absolute number of bytes not an increment.   The fact that it can't shrink is a current implementation detail isn't it?  With wasm IIUC there it is never possible to shrink memory I doubt that will ever change.

You are renaming the parameter to `requested_growth_bytes` but I don't see the implementation changing.

I think the terms `heap` and `buffer` and `realloc` are all a little confusing. IIUC what is really being resized here is the webassembly memory which hold static data, stack and heap. 

Maybe we can come up with better names for these and some documentation.   Do we *really* need two different memory resize functions?  If we do shouldn't they be called something like:

- `emscripten_memory_grow`
- `emscripten_memory_resize`

?

Reply:
> You are renaming the parameter to `requested_growth_bytes` but I don't see the implementation changing.

Thanks, good catch, when doing a documentation pass I misremembered the function argument behavior. Updated.

> Do we _really_ need two different memory resize functions?

Yes - while the automatic overgrowing is useful for some applications, for others it gets in the way if they don't want to do sbrk() or similar. I like the names `emscripten_memory_grow()` and `emscripten_memory_resize`, let's go with those. @kripken wdyt?

Reply:
Actually, on another thought, I don't want to grow this PR and start renaming functions here, that should be for later. Note that the implementation here does not add any new heap resize/grow functionality, but just exposes the function that has already existed for a long time.